### PR TITLE
feat(resharding): fix 4 places related to outgoing receipts

### PR DESCRIPTION
**Pull request by Longarithm**
I'm not doing it separately because for my test setups partial fixes didn't work, and we are shipping soon... So I'll explain everything here.

1. Filtering receipts

On the shard split boundary, we must filter outgoing receipts from parent chunks, because not all of them actually target the first chunk in our shard. **But the chunk proof we send to chunk validators must include all receipts.** That's why we need to add awful parameter `filter_receipts` to skip filtering when we generate proof, and enable filtering when we apply chunk.
Also filtering was completely missing on chunk validation; added that.

2. State dumper fix

`verify_path` assertion was failing. 

<details>
<summary>[OUTDATED]</summary>
`compute_state_response_header` wasn't fully reflecting resharding logic after moving sync hash to current epoch.
First, last chunk' shard layout corresponds to the same epoch id, not the previous one.
Second, on aggregating incoming receipts, shard layout actually **may** change, so we need to maintain target shard id properly.
</details>

CurrentEpochStateSync provides sync_hash for which there are 2 chunks for each shard. However, state sync code goes 1 block back so it gives only 1 chunk. To get incoming receipts for this chunk, we will cross epoch and resharding boundary. So I move sync_hash one block forward for now.

These two cases are caught by `test_resharding_v3_shard_shuffling_intense`, which creates nontrivial outgoing receipt distributions on resharding boundary. The next two cases are more insidious and caught by patch #12597. `test_resharding_v3_outgoing_receipts_from_splitted_shard` and similar tests were useful.
I tried to make a proper test but I think I need to enable single shard tracking everywhere. For that I tried to add rpc node tracking all shards but it wasn't simple because we have assumptions on what is rpc node account id.

3. Reassignment

Once I understood what `ChainStore::reassign_outgoing_receipts_for_resharding` is for, I understood as well that it must be called for chunk validators. It's better to read its definition on review. Without it I was getting `InvalidReceiptsProof` for **one of the parent shards**.

4. Cache by target shard id

`MainStateTransitionCache` must cache results by child shard id, not parent, otherwise there is a collision. For shard merging we probably need to key entries by both shard ids.

Maybe we can un-ignore some of tests after that, I didn't check.
It also uses #12586.

### CODE REVIEW

File: chain/chain/src/store/mod.rs [hunk 1]
```diff
@@ -283,10 +272,24 @@ pub trait ChainStoreAccess {
                     // incoming receipts. It would be nicer to explicitly check
                     // that condition rather than relying on errors when reading
                     // from the db.
-                    ret.push(ReceiptProofResponse(current_block_hash, Arc::new(vec![])));
+                    Arc::new(vec![])
                 }
-            }
+            };
+
+            // If the shard layout changed we need to filter receipts to
```
Comment (Trisfald) [inline chain/chain/src/store/mod.rs:279]: So now, is it more like "in some places" we need to filter receipts and "in other places" (state witness) we don't 😄 ? 
 
🤯 🤯

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -2594,19 +2592,22 @@ impl Chain {
...
 
         // Collecting the `prev` state.
         let (prev_chunk_header, prev_chunk_proof, prev_chunk_height_included) = match self
-            .get_block(block_header.prev_hash())
+            .get_block(chunk_block_header.prev_hash())
         {
             Ok(prev_block) => {
+                let (prev_shard_id, prev_shard_index) = self
+                    .epoch_manager
+                    .get_prev_shard_id(chunk_block_header.prev_hash(), shard_id)?;
```
Comment (wacban) [inline chain/chain/src/chain.rs:2606]: It took me a while to convince myself that this is indeed prev and not prev prev.

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -2594,19 +2592,22 @@ impl Chain {
...
                 let prev_chunk_header = prev_block
                     .chunks()
                     .get(prev_shard_index)
-                    .ok_or(Error::InvalidShardId(shard_id))?
+                    .ok_or(Error::InvalidShardId(prev_shard_id))?
```
Comment (wacban) [inline chain/chain/src/chain.rs:2610]: nice, my bad

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -2641,14 +2642,18 @@ impl Chain {
...
         // new epoch.
+        // TODO: why we collect incoming receipts from `sync_hash`, not
+        // `chunk_block_header`?
         let incoming_receipts_proofs = self.chain_store.get_incoming_receipts_for_shard(
             self.epoch_manager.as_ref(),
             shard_id,
             &shard_layout,
             sync_hash,
             prev_chunk_height_included,
+            false,
```
Comment (wacban) [inline chain/chain/src/chain.rs:2653]: nit: bools are a bit cryptic, consider using an enum

File: chain/chain/src/chain.rs [hunk 5]
```diff
@@ -2672,8 +2677,10 @@ impl Chain {
             for receipt_proof in receipt_proofs.iter() {
                 let ReceiptProof(receipts, shard_proof) = receipt_proof;
                 let ShardProof { from_shard_id, to_shard_id: _, proof } = shard_proof;
-                let receipts_hash = CryptoHash::hash_borsh(ReceiptList(shard_id, receipts));
-                let from_shard_index = prev_shard_layout.get_shard_index(*from_shard_id)?;
+                let receipts_hash = CryptoHash::hash_borsh(ReceiptList(target_shard_id, receipts));
+                let shard_layout =
+                    self.epoch_manager.get_shard_layout(block.header().epoch_id())?;
```
Comment (wacban) [inline chain/chain/src/chain.rs:2682]: nit: move before the loop

File: chain/chain/src/chain.rs [hunk 6]
```diff
@@ -2672,8 +2677,10 @@ impl Chain {
...
+                let from_shard_index = shard_layout.get_shard_index(*from_shard_id)?;
```
Comment (wacban) [inline chain/chain/src/chain.rs:2683]: I'm tempted to implement a debug assert feature where we add an ephemeral `version` to shard layout and shard id and we assert that shard id can only be used with shard layout with same version. It would require a lot of refactoring though :(

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 7]
```diff
@@ -166,27 +169,29 @@ fn get_state_witness_block_range(
...
 
     let initial_prev_hash = *state_witness.chunk_header.prev_block_hash();
     let initial_prev_block = store.get_block(&initial_prev_hash)?;
     let initial_shard_layout =
         epoch_manager.get_shard_layout_from_prev_block(&initial_prev_hash)?;
     let initial_shard_id = state_witness.chunk_header.shard_id();
-    let initial_shard_index = initial_shard_layout.get_shard_index(initial_shard_id)?;
+    // Check that shard id is present in current epoch.
+    // TODO: consider more proper way to validate this.
+    let _ = initial_shard_layout.get_shard_index(initial_shard_id)?;
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:187]: hehe nice

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 8]
```diff
@@ -205,6 +210,8 @@ fn get_state_witness_block_range(
         }
         let (prev_shard_id, prev_shard_index) =
             epoch_manager.get_prev_shard_id(prev_hash, position.shard_id)?;
+        let prev_shard_layout =
+            epoch_manager.get_shard_layout(&epoch_manager.get_epoch_id(prev_hash)?)?;
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:214]: I'm not convinced that those are matching, the get_prev_shard_id logic is non-trivial. Perhaps you can refactor it to have the `get_prev_shard_id` return all `prev_shard_id`, `prev_shard_index` and `prev_shard_layout`?

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 9]
```diff
@@ -580,16 +597,22 @@ pub fn validate_chunk_state_witness(
         .with_label_values(&[&state_witness.chunk_header.shard_id().to_string()])
         .start_timer();
     let span = tracing::debug_span!(target: "client", "validate_chunk_state_witness").entered();
+    let witness_shard_layout = epoch_manager.get_shard_layout(&state_witness.epoch_id)?;
+    let witness_chunk_shard_id = state_witness.chunk_header.shard_id();
+    let witness_chunk_shard_uid =
+        epoch_manager.shard_id_to_uid(witness_chunk_shard_id, &state_witness.epoch_id)?;
     let block_hash = pre_validation_output.main_transition_params.block_hash();
     let epoch_id = epoch_manager.get_epoch_id(&block_hash)?;
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:605]: It's a bit annoying that we have two different epoch ids - the `state_witness.epoch_id` and this one. I don't have a good suggestion, just ranting.

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 10]
```diff
@@ -622,14 +645,23 @@ pub fn validate_chunk_state_witness(
 
     // Compute receipt hashes here to avoid copying receipts
     let outgoing_receipts_hashes = {
-        let shard_layout = epoch_manager
-            .get_shard_layout_from_prev_block(state_witness.chunk_header.prev_block_hash())?;
-        Chain::build_receipts_hashes(&outgoing_receipts, &shard_layout)
+        let chunk_epoch_id = epoch_manager.get_epoch_id(&block_hash)?;
+        let chunk_shard_layout = epoch_manager.get_shard_layout(&chunk_epoch_id)?;
+        if chunk_shard_layout != witness_shard_layout {
+            ChainStore::reassign_outgoing_receipts_for_resharding(
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:651]: Nice! Good find!

File: chain/chain/src/store/mod.rs [hunk 11]
```diff
@@ -211,6 +211,7 @@ pub trait ChainStoreAccess {
         target_shard_layout: &ShardLayout,
         block_hash: CryptoHash,
         last_chunk_height_included: BlockHeight,
+        filter_receipts: bool,
```
Comment (wacban) [inline chain/chain/src/store/mod.rs:214]: Can you add a comment explaining the circumstances where this is set to true or false?

File: chain/chain/src/store/mod.rs [hunk 12]
```diff
@@ -578,7 +581,7 @@ impl ChainStore {
     }
 
     /// TODO validate if this logic works for Resharding V3.
-    fn reassign_outgoing_receipts_for_resharding(
+    pub fn reassign_outgoing_receipts_for_resharding(
```
Comment (wacban) [inline chain/chain/src/store/mod.rs:584]: Is the TODO done? Can you remove it if so?

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 13]
```diff
@@ -467,6 +470,55 @@ fn check_promise_yield_receipts_exist_in_memtrie(
...
+            let client_actor = &mut test_loop_data.get_mut(&client_handle);
+            let tip = client_actor.client.chain.head().unwrap();
+
+            // Run this action only once at every block height.
+            if latest_height.get() == tip.height {
+                return;
+            }
+            latest_height.set(tip.height);
+
+            let mut rng = rand::thread_rng();
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:493]: This is debatable, but it's more common in our tests to use seedable rng to make it all deterministic and reproducible. If you want the extra randomness you can add multiple tests with different seeds.

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 14]
```diff
@@ -467,6 +470,55 @@ fn check_promise_yield_receipts_exist_in_memtrie(
...
+            for _ in 0..NUM_TRANSFERS_PER_BLOCK {
+                let sender = account_ids.choose(&mut rng).unwrap().clone();
+                let receiver = account_ids.choose(&mut rng).unwrap().clone();
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:496]: ❤️ This is awesome, we totally need tests with some random traffic across all shards.

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 15]
```diff
@@ -1370,6 +1422,26 @@ fn test_resharding_v3_shard_shuffling() {
...
+fn test_resharding_v3_shard_shuffling_intense() {
+    let chunk_ranges_to_drop = HashMap::from([
+        (ShardUId { shard_id: 0, version: 3 }, -1..2),
+        (ShardUId { shard_id: 1, version: 3 }, -3..0),
+        (ShardUId { shard_id: 2, version: 3 }, -3..0),
+        (ShardUId { shard_id: 3, version: 3 }, 0..3),
+        (ShardUId { shard_id: 4, version: 3 }, 0..1),
+    ]);
+    let mut params = TestReshardingParameters::new()
+        .base_shard_layout_version(1)
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1435]: I would advice either:
* use the default
* make this test parametrized by the version and have one test for each

File: integration-tests/src/test_loop/utils/transactions.rs [hunk 16]
```diff
@@ -56,12 +57,15 @@ pub(crate) fn get_anchor_hash(clients: &[&Client]) -> CryptoHash {
 }
 
 /// Get next available nonce for the account's public key.
-pub fn get_next_nonce(env: &mut TestLoopEnv, account_id: &AccountId) -> u64 {
+pub fn get_next_nonce(
+    test_loop_data: &TestLoopData,
+    node_datas: &[TestData],
```
Comment (wacban) [inline integration-tests/src/test_loop/utils/transactions.rs:62]: Just out of curiosity why this change? Did you have some trouble with mut?

**Review by wacban — APPROVED** [review]
LGTM, great stuff, it's really unblocking us! 

Please see the comment about `get_prev_shard_id` this is the only serious one here.

File: integration-tests/src/test_loop/tests/resharding_v3.rs (hunk 13 above)
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:493]: Agree, perhaps it's better to find a seed which triggers the error (without the fix) and always use that one.

File: chain/chain/src/stateless_validation/chunk_validation.rs (hunk 8 above)
Comment (Longarithm) [reply chain/chain/src/stateless_validation/chunk_validation.rs:214]: I'm not happy about extending return result even further, but... works for now.

File: chain/chain/src/chain.rs (hunk 2 above)
Comment (Longarithm) [reply chain/chain/src/chain.rs:2606]: You mean that `get_prev_shard_id` accepts `prev_hash`? Yeah... I'll go ahead and rename with `get_prev_shard_id_from_prev_hash`.

**Review by marcelo-gonzalez — COMMENTED** [review]
Looks sane to me, thanks for fixing it! I'm just going to try check the receipts logic change in `compute_state_header()` to make sure it still works with `set_state_finalize()`, because it feels tricky to me.

File: integration-tests/src/test_loop/tests/resharding_v3.rs (hunk 13 above)
Comment (Longarithm) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:493]: But is it more common? Search by `rand::thread_rng()` gives more results than by `SeedableRng`.
And fixing the seed defeats the purpose of random.
If we fix the seed, we should as well find the specific pattern and hardcode the pattern instead of keeping randomness.
Without fixed seed, the test will keep finding patterns for us in order to catch other bugs.
Then it turns into a question - what is better: CI suddenly finding a failure for us or CI being more stable and reproducible? I think the first is better.

What I probably should already do is to make 10 iterations of this test to reduce flakiness. But I need more evidence why seedable random is better.

File: chain/chain/src/chain.rs (hunk 2 above)
Comment (Longarithm) [reply chain/chain/src/chain.rs:2606]: Lol. Yeah, I agree.

We really really need to make this logic simpler and try to extract common logic. This block iteration and receipts aggregation is kinda duplicated in many places; and where we try to use common functions, there are usually too many parameters. I had some trouble with `get_state_witness_block_range`.
I also think it's worthy to show my thinking process.

First - I didn't know (forgot) that new sync hash guarantees prev_shard_id = shard_id, so I added this.
Second - I understood your comment, but thought that we still need to call get_prev_shard_id to support old sync hash.
Third - I realised that we are likely to enable CurrentEpochStateSync not later than SimpleNightshadeV4, so we don't need to call it.

But that's too much thinking which justifies the need for refactoring. Anyway, I guess yeah, the change makes logic slightly better.

File: chain/chain/src/chain.rs [hunk 17]
```diff
@@ -2573,14 +2573,12 @@ impl Chain {
         let sync_prev_block = self.get_block(sync_block_header.prev_hash())?;
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply chain/chain/src/chain.rs]: I'll skip for now because as Marcelo showed that this code part shouldn't be changed, so I reverted it. The fix is only related to receipts aggregation

File: integration-tests/src/test_loop/tests/resharding_v3.rs (hunk 13 above)
Comment (wacban) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:493]: Ah perhaps I just saw more of it than the pure random. 

I think we're mixing two different types of testing. The first one is the unit / integration test where determinism should be preferred in my opinion. The second is randomized / fuzzing tests where we want to cover as many cases as possible. Given we don't have a good infra for fuzzing tests I'm fine with keeping the integration tests random. 

I do have a small request though - can you pick a seed at random, log it and then use the deterministic rng? This way if the test breaks we will be easily able to reproduce it still - by hardcoding the seed to the logged one.

File: chain/chain/src/chain.rs [hunk 18]
```diff
@@ -2573,14 +2573,12 @@ impl Chain {
...
-        let prev_epoch_id = sync_prev_block.header().epoch_id();
-        let prev_shard_layout = self.epoch_manager.get_shard_layout(&prev_epoch_id)?;
-        let prev_shard_index = prev_shard_layout.get_shard_index(shard_id)?;
+        let shard_index = shard_layout.get_shard_index(shard_id)?;
 
         // Chunk header here is the same chunk header as at the `current` height.
         let sync_prev_hash = sync_prev_block.hash();
         let chunks = sync_prev_block.chunks();
-        let chunk_header = chunks.get(prev_shard_index).ok_or(Error::InvalidShardId(shard_id))?;
+        let chunk_header = chunks.get(shard_index).ok_or(Error::InvalidShardId(shard_id))?;
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply chain/chain/src/chain.rs]: + renamed parameters to sync_prev_shard_index, etc.

File: chain/chain/src/state_sync.rs [hunk 19]
```diff
@@ -56,23 +56,20 @@ fn save_epoch_new_chunks<T: ChainStoreAccess>(
...
-    for (num_new_chunks, new_chunk) in num_new_chunks.iter_mut().zip(header.chunk_mask().iter()) {
-        // Only need to reach 2, so don't bother adding more than that
-        if *new_chunk && *num_new_chunks < 2 {
-            *num_new_chunks += 1;
-        }
-        if *num_new_chunks < 2 {
-            done = false;
-        }
-    }
+    let done = num_new_chunks.iter().all(|num_chunks| *num_chunks >= 2);
```
Comment (marcelo-gonzalez) [inline chain/chain/src/state_sync.rs:59]: Ahh was this a bug before too? So this change will push out the sync hash one more block (with no missed chunks, 4th block of the epoch instead of the 3rd). Was the previous one incorrect somewhere? The logic in that state sync header generation is a bit tricky, could def have missed something there

Comment (marcelo-gonzalez) [reply chain/chain/src/state_sync.rs:59]: Oh nvm I see your msg on zulip, yea this is a good fix, thanks!

I think this was never caught because the `state_sync.rs` tests [set up skipped chunks](https://github.com/near/nearcore/blob/97f69fe08ee4744072374d2a96cb8db9da31afcf/integration-tests/src/test_loop/tests/state_sync.rs#L143), but this wont be hit there bc theres no resharding. And the resharding shard shuffling test doesnt skip chunks. So maybe would be good to add some resharding tests w/ `drop_chunks_by_height()` or something

**Comment by marcelo-gonzalez** [comment]
nice, LGTM on the state sync stuff, although I am not super familiar w the chunk state witness code

File: chain/chain/src/chain.rs [hunk 20]
```diff
@@ -2689,6 +2696,9 @@ impl Chain {
                 ));
             }
             root_proofs.push(root_proofs_cur);
+            (target_shard_id, _) = self
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply chain/chain/src/chain.rs]: Because `target_shard_id` was already known for `block_hash`, and this function moves shard id 1 block back, so it must happen in the end. (Anyway I reverted changes in the file)

File: chain/chain/src/store/mod.rs (hunk 1 above)
Comment (Longarithm) [reply chain/chain/src/store/mod.rs:279]: removed the comment

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 16 above)
Comment (Longarithm) [reply integration-tests/src/test_loop/utils/transactions.rs:63]: It's because I didn't have `TestLoopEnv` for event handling, and I also removed  `mut` because it is read-only call.
